sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionEvalHelperSuite.scala

```diff
@@ -35,6 +36,13 @@ class ExpressionEvalHelperSuite extends SparkFunSuite with ExpressionEvalHelper
     val e = intercept[RuntimeException] { checkEvaluation(BadCodegenExpression(), 10) }
     assert(e.getMessage.contains("some_variable"))
   }
+
+  test("SPARK-25388: checkEvaluation should fail if nullable in DataType is incorrect") {
+    val e = intercept[RuntimeException] {
+      checkEvaluation(MapIncorrectDataTypeExpression(), Map(3 -> 7, 6 -> null))
```

Comment:
Here is an output. This is because the test correctly detects a failure even in codegen-off mode, too.

```
"Incorrect evaluation (codegen off): mapincorrectdatatypeexpression(), actual: keys: [3,6], values: [7,null], expected: keys: [3,6], values: [7,-1]" did not contain "Incorrect evaluation in unsafe mode"
ScalaTestFailureLocation: org.apache.spark.sql.catalyst.expressions.ExpressionEvalHelperSuite$$anonfun$4 at (ExpressionEvalHelperSuite.scala:44)
org.scalatest.exceptions.TestFailedException: "Incorrect evaluation (codegen off): mapincorrectdatatypeexpression(), actual: keys: [3,6], values: [7,null], expected: keys: [3,6], values: [7,-1]" did not contain "Incorrect evaluation in unsafe mode"
	at org.scalatest.Assertions$class.newAssertionFailedException(Assertions.scala:528)
...
```

Reply:
I see, so the example above passes in codegen off and fails with codegen on with this fix, while using `Map(3 -> 7, 6 -> -1)` passes codegen on and fails codegen off, am I right?

What I am thinking about (but I have not yet found a working implementation) is: since the problem arise when we say we expect `null` in a non-nullable datatype, can we add such a check? I mean, instead of pretending the expected value to be nullable, can't we add a check in case it is not nullable for being sure that it does not contain `null`? I think it would be better, because we would be able to distinguish a failure caused by a bad test, ie. a test written wrongly, from a UT failure caused by a bug in what we are testing. What do you think?

Reply:
Here is summary.
* `Map(3 -> 7, 6 -> null)` passes in codegen off and fails in codegen on with this fix
* `Map(3 -> 7, 6 -> -1)` fails in codegen off and fails in codegen on 

Would it be possible to share examples of two cases that you think we would be able to distinguish?

Reply:
yes, thanks for the summary, it states more clearly what I thought.

My point is that this fix works properly only when we test both codegen on and off, but it would fail to detect the error condition it claims to fix if only one of them (for any reason) is tested. So I am wondering if it is possible to perform a check on the expected value, instead of this fix. Something like:
```
assert(containsNull(expected) && isNullable(expression.dataType))
```
where `containsNull` and `isNullable` have to be defined properly. In this way we should fail properly independently from whether codegen is on or not. And we can also give a more clear hint in the error message about the problem being most likely a bad UT.

Reply:
Would it be possible to let us know how can this assertion distinguish the following two cases?
* a failure caused by a bad test, ie. a test written wrongly
* a UT failure caused by a bug in what we are testing

Reply:
Well, actually I was not clear about this, sorry for that. I meant we can separate from:
 - a failure caused by a wrong result returned;
 - a failure caused by a wrong result type returned (ie. the expected values contains NULL while the expression return data type is not nullable - this case can happen both in case the UT is bad and in case the return type of the expression is wrong);

Reply:
I meant how your proposed assert can distinguish the two cases that you want to distinguish at [here](https://github.com/apache/spark/pull/22375#discussion_r217495874).
1. a failure caused by a bad test, ie. a test written wrongly
2. a UT failure caused by a bug in what we are testing

As you said [here](https://github.com/apache/spark/pull/22375#discussion_r218362156), the assert may detect both cases 1. and 2..

Reply:
Sorry, you're right, I have not been precise in my previous statements so they are confusing. I'll try to elaborate more clearly.

IIUC what you are proposing now is for detecting the case when the output datatype of an expression is not nullable and the expected output instead contains a null. The way you're trying to achieve this is to make always nullable the expected output, so that we get a null as output both for codegen on and codegen off (while before this change for codegen on we would get as expected output a wrong value, which is the default value for that data type).

The problems I see on the approach above are:
 - if you test only codegen on or codegen off, you might still be missing to cover some cases (eg. see https://github.com/apache/spark/pull/22375#discussion_r218085246, with codegen off we can have null as output for a non-nullable expression...);
 - The failure you get is that the expected output is different from the the evaluated one, so this case is treated as any other failure in the expression evaluation.

With something like this:
```
assert(containsNull(expected) && isNullable(expression.dataType))
```
I think we could solve the 2 problems above as:
 - this check would fail in both modes;
 - we can differentiate between a test failing because of a wrong output and this new case which covers: a bad written UT; an expression returning wrongly a non-nullable datatype instead of a nullable one when returning null.

Hope now it is more clear. Thanks.

Reply:
The your first is correct since this patch addresses only codegen-on case. We can add another code to address codegen-off case.
Regarding the your second point, have we ever distingished a wrong output from a bad written UT when we defect the difference between `expression` and `expected`. I think that the distinguishment is nice to have, but not mandatory to have.


I have one question about your approach:
```
assert(containsNull(expected) && isNullable(expression.dataType))
```
Since the above two conditions evaluates `expected` and `expression` independently, how this works for the following case? I think that the assertion would be passed
```
expression:
dataType = StructType(ArrayType(IntegerType, false), ArrayType(IntegerType, true))
Struct(Array(0, null), Array(1, 0))

expected:
Struct(Array(0, 0), Array(1, null))
```

Reply:
Yes, you're right, my suggestion doesn't work in a case like that, sorry. We would need to make it checking recursively. But I think you got the idea of what I am proposing here. Thanks.

Reply:
Even if we make it checking recursively, I think that this case cannot be detected. This is because the mismatch occurs in the different recursive path.

Would it be possible to share the case where we distingished a wrong output from a bad written UT in other places, as you proposed?

Reply:
I may not still understand your motivation correctly. What is the motivation to introduce this assertion?

Reply:
The motivations are the 2 mentioned above. Basically, I am proposing the same suggestion @cloud-fan has just commented [here](https://github.com/apache/spark/pull/22375#discussion_r219452615)

Reply:
With some hints from @ueshin, this PR implemented the check of `null` value with nullable bit in `checkResult()`.